pallets/inflation/src/lib.rs

```diff
@@ -0,0 +1,604 @@
+            let dapps_emission = params.dapps_part * max_emission;
+            let base_stakers_emission = params.base_stakers_part * max_emission;
+            let adjustable_stakers_emission = params.adjustable_stakers_part * max_emission;
+            let bonus_emission = params.bonus_part * max_emission;
+
+            // 3. Calculate concrete rewards per block, era or period
+
+            // 3.0 Convert all 'per cycle' values to the correct type (Balance).
+            // Also include a safety check that none of the values is zero since this would cause a division by zero.
+            // The configuration & integration tests must ensure this never happens, so the following code is just an additional safety measure.
+            let blocks_per_cycle = match T::CycleConfiguration::blocks_per_cycle() {
+                0 => Balance::MAX,
+                blocks_per_cycle => Balance::from(blocks_per_cycle),
+            };
+
+            let build_and_earn_eras_per_cycle =
+                match T::CycleConfiguration::build_and_earn_eras_per_cycle() {
+                    0 => Balance::MAX,
+                    build_and_earn_eras_per_cycle => Balance::from(build_and_earn_eras_per_cycle),
+                };
+
+            let periods_per_cycle = match T::CycleConfiguration::periods_per_cycle() {
+                0 => Balance::MAX,
+                periods_per_cycle => Balance::from(periods_per_cycle),
+            };
+
+            // 3.1. Collator & Treausry rewards per block
+            let collator_reward_per_block = collators_emission / blocks_per_cycle;
+            let treasury_reward_per_block = treasury_emission / blocks_per_cycle;
+
+            // 3.2. dApp reward pool per era
+            let dapp_reward_pool_per_era = dapps_emission / build_and_earn_eras_per_cycle;
+
+            // 3.3. Staking reward pools per era
+            let base_staker_reward_pool_per_era =
+                base_stakers_emission / build_and_earn_eras_per_cycle;
+            let adjustable_staker_reward_pool_per_era =
+                adjustable_stakers_emission / build_and_earn_eras_per_cycle;
+
+            // 3.4. Bonus reward pool per period
+            let bonus_reward_pool_per_period = bonus_emission / periods_per_cycle;
+
+            // 4. Block at which the inflation must be recalculated.
+            let recalculation_block = now.saturating_add(T::CycleConfiguration::blocks_per_cycle());
+
+            // 5. Return calculated values
+            (
+                max_emission,
+                InflationConfiguration {
+                    recalculation_block,
+                    collator_reward_per_block,
+                    treasury_reward_per_block,
+                    dapp_reward_pool_per_era,
+                    base_staker_reward_pool_per_era,
+                    adjustable_staker_reward_pool_per_era,
+                    bonus_reward_pool_per_period,
+                    ideal_staking_rate: params.ideal_staking_rate,
```

Comment:
Do we need `ideal_staking_rate` in both `InflationConfiguration` and `InflationParameters`? It seems duplicated to me.

Reply:
I believe we do.

This is a static parameter, as well as active config value.
Based on:
* params
* total issuance
new config is derived. It seems cleaner to me to have the two separate.

However, the main reason why I decided to do this is to have all the necessary information
for reward payout in a single struct.